File: Simulation_Conley.R
##### This file implements the simulation setting proposed in Conley et al (2008) #####

source("MartingalePosteriorGMM.R") # load the Martingale posterior function
library(bayesm) # The bayesm package is needed for the competing methods

## function to generate the data ##
generate_data = function(n, s = 1, beta = 1){
  alpha = 0
  gamma = 0
  delta = rep(s, 10)
  Sigma = matrix(c(1, 0.6, 0.6, 1), ncol = 2)
  
  u = exp(MASS::mvrnorm(n, c(0, 0), 0.6 * Sigma))
  z = matrix(runif(10 * n), ncol = 10, nrow = n)
  x = gamma + z %*% delta + u[, 1]
  y = alpha + beta * x + u[, 2]
  return(list(y = y[, 1], x = x, z = z))
}

## functions to compute the performance criteria
mae = function(beta_estimates, true_beta = 1){
  return(median(abs(beta_estimates - true_beta)))
}
coverage = function(ci_lower, ci_upper, true_beta = 1){
  return(mean((ci_lower < true_beta) & (ci_upper > true_beta)))
}

## function that extracts the quantities of interest from the fit objects
compute_quantities = function(fit_list){
  # create list of the posterior samples
  post_list = list(
    sapply(fit_list$MP_DDP, "[[", 2),
    sapply(fit_list$MP_LM, "[[", 2),
    as.numeric(fit_list$Bayes_IV$betadraw),
    as.numeric(fit_list$Bayes_IV_DP$betadraw)
  )
  
  # return quantities of interest (point estimates and credible/confidence interval)
  return(list(
    point_estimates = c(sapply(post_list, median), fit_list$TSLS$coef[2]),
    ci_lower = c(sapply(post_list, quantile, 0.025), fit_list$TSLS$ci_lower),
    ci_upper = c(sapply(post_list, quantile, 0.975), fit_list$TSLS$ci_upper)
  ))
  
}

## wrapper function that runs the simulation
run_simulation = function(s, M = 100, n = 100, beta = 1, N = 500, B = 500){
  methods = c("MPIV (DDP)", "MPIV (LM)", "Bayes IV", "Bayes IV (DP)", "TSLS")
  
  # create storage objects
  point_estimates = matrix(NA, ncol = M, nrow = length(methods))
  ci_lower = matrix(NA, ncol = M, nrow = length(methods))
  ci_upper = matrix(NA, ncol = M, nrow = length(methods))
  
  # loop over M iterations
  pb <- txtProgressBar(min = 0, max = M, style = 3)
  for (j in 1:M) {
    # generate data
    d = generate_data(n, s = s)
    
    # fit models
    fit_list = list(
      "MP_DDP" = martingale_posterior(d$y, d$x, d$z, N = N, B = B, type = "DDP"),
      "MP_LM" = martingale_posterior(d$y, d$x, d$z, N = N, B = B, type = "LM"),
      "Bayes_IV" = bayesm::rivGibbs(
        list(y = d$y, x = d$x[, 1], w = matrix(1, ncol = 1, nrow = n), z = cbind(1, d$z)),
        Mcmc = list(R = 2*B, keep = 2, nprint = 0),
      ),
      "Bayes_IV_DP" = bayesm::rivDP(
        list(y = d$y, x = d$x[, 1], z = d$z),
        Mcmc = list(R = 2*B, keep = 2, nprint = 0),
      ),
      "TSLS" = tsls(d$y, d$x, d$z, ci = TRUE)
    )
    
    quantities = compute_quantities(fit_list)
    
    # extract quantities of interest
    point_estimates[, j] = quantities$point_estimates
    ci_lower[, j] = quantities$ci_lower
    ci_upper[, j] = quantities$ci_upper
    
    # Update the progress bar
    setTxtProgressBar(pb, j)
  }
  
  # compute performance measures
  coverage_results = numeric(length(methods))
  median_interval_length = numeric(length(methods))
  for (i in 1:length(methods)) {
    coverage_results[i] = coverage(ci_lower[i, ], ci_upper[i, ])
    median_interval_length[i] = median(ci_upper[i, ] - ci_lower[i, ])
  }
  
  res = cbind(
    "MAE" = apply(point_estimates, 1, mae),
    "Coverage" = coverage_results,
    "MIL" = median_interval_length
  )
  rownames(res) = methods
  return(res)
}

# Run the simulation
# We vary s between 0.5, 1, and 1.5 corresponding to weak, moderate and strong instruments
ss = c(0.5, 1, 1.5)
result = lapply(ss, run_simulation)
setNames(result, paste0("s = ", ss))
saveRDS(result, file = "Results_Conley.RDS")

File: simulation_conley.jl


include("MartingalePosterior.jl")

using Base.Threads # for parallelisation
using ProgressMeter
using JLD2
using RCall

# Data generating function (DGP from Conley et. al., 2008)
function generate_data(n::Int, s::Real = 1, beta::Real = 1)
    alpha = 0.0
    gamma = 0.0
    delta = fill(s, 10)
    Sigma = [1.0 0.6; 0.6 1.0]

    mvnorm = MvNormal(zeros(2), 0.6 * Sigma)
    u = exp.(rand(mvnorm, n)')
    #u = rand(mvnorm, n)'

    z = rand(Uniform(0, 1), n, 10)
    x = gamma .+ z * delta .+ u[:, 1]
    y = alpha .+ beta * x .+ u[:, 2]

    return (y = y, x = x, z = z)
end

# Implement the method of Conley at. al. (2008)
function iv_conley(y, x, z; B = 100)
    @rput y x z B
    R"""
    fit = bayesm::rivDP(
        list(y = y, x = x, z = z),
        Mcmc = list(R = 2*B, keep = 2, nprint = 0),
    )
    betas = as.numeric(fit$betadraw)
    """
    @rget betas
    return betas
end

function iv_rossi(y, x, z; B = 100)
    @rput y x z B
    R"""
    fit = bayesm::rivGibbs(
        list(y = y, x = x, w = matrix(1, ncol = 1, nrow = length(y)), z = cbind(1, z)),
        Mcmc = list(R = 2*B, keep = 2, nprint = 0),
      )
    betas = as.numeric(fit$betadraw)
    """
    @rget betas
    return betas
end


# function that computes the performance criteria
function performance_measures(posterior_sample::Vector{Float64}, true_value::Float64)
    post_median = median(posterior_sample)
    lower, upper = quantile(posterior_sample, [0.025, 0.975])
    return (
        absolute_error = abs( post_median - true_value),
        coverage_flag = lower <= true_value <= upper,
        interval_length = upper - lower
    )
end

# Wrapper function that runs the simulation
function run_simulation(s::Float64; M::Int = 100, n::Int = 100, B::Int = 100, true_value::Float64 = 1.0)
    # Preallocate arrays
    methods = ["Martingale Posterior (MF)", "Bayes IV", "Bayes IV (DP)"]
    abs_errors = zeros(length(methods), M)
    coverage_flags = falses(length(methods), M)
    interval_lengths = zeros(length(methods), M)

    #Threads.@threads for i in 1:M
    @showprogress for i in 1:M
        # Simulate data
        y, x, z = generate_data(n, s, true_value)

        # Get posterior samples
        mp_fit = martingale_posterior(y, x, z; B = B)
        mp = getindex.(mp_fit, 2)
        iv = iv_rossi(y, x, z; B = B)
        ivdp = iv_conley(y, x, z; B = B)

        # compute performance criteria
        abs_errors[1, i], coverage_flags[1, i], interval_lengths[1, i] = performance_measures(mp, true_value)
        abs_errors[2, i], coverage_flags[2, i], interval_lengths[2, i] = performance_measures(iv, true_value)
        abs_errors[3, i], coverage_flags[3, i], interval_lengths[3, i] = performance_measures(ivdp, true_value)
    end

    # Compute performance measures
    mae = median(abs_errors; dims = 2)
    coverage = mean(coverage_flags; dims = 2)
    median_interval_length = mean(interval_lengths; dims = 2)

    return (MAE = mae, Coverage = coverage, MIL = median_interval_length)
end

ss = [0.5, 1.0, 1.5]
result = map(run_simulation, ss)
print(result)
save("Results_Conley.jld2", Dict("s=0.5" => result[1], "s=1" => result[2], "s=1.5" => result[3]))

